src/generators/ast.ts

```diff
@@ -198,10 +204,97 @@ function copyAdvancedKeywords(from: JSONSchema7, to: Schema): void {
   }
 }
 
+type ResolvedTypes = {
+  [key: string]: Schema;
+};
+
+let definedTypes: ResolvedTypes = {};
+
+export interface CustomType {
+  name: string;
+  schema: Schema;
+}
+
+export const customTypesByEvent: Record<string, CustomType[]> = {};
+
+export function extractCustomTypes(schema: JSONSchema7, eventName?: string): void {
+  definedTypes = {};
+
+  // Early return if no schema definitions or event name
+  if (!schema.$defs || !eventName) {
+    return;
+  }
+
+  const customTypes: CustomType[] = [];
+
+  // First pass: extract all custom types
+  for (const [name, defSchema] of Object.entries(schema.$defs)) {
+    // Skip boolean schemas
+    if (typeof defSchema === 'boolean') {
+      continue;
+    }
+
+    const customTypeSchema: JSONSchema7 = {
+      ...defSchema,
+      title: name,
+      description: defSchema.description || `Custom type for ${eventName}`,
+    };
+
+    definedTypes[`#/$defs/${name}`] = customTypeSchema as any;
+    definedTypes[name] = customTypeSchema as any;
+
+    customTypes.push({
+      name,
+      schema: customTypeSchema as any,
+    });
+  }
+
+  // Early return if no custom types found
+  if (customTypes.length === 0) {
+    return;
+  }
+
+  // Second pass: parse all custom types now that they're all defined
+  for (const customType of customTypes) {
+    const parsedSchema = parse(customType.schema as unknown as JSONSchema7, customType.name);
+    definedTypes[`#/$defs/${customType.name}`] = parsedSchema;
+    definedTypes[customType.name] = parsedSchema;
+    customType.schema = parsedSchema;
+  }
+
+  customTypesByEvent[eventName] = customTypes;
+}
+
 // parse transforms a JSON Schema into a standardized Schema.
 export function parse(raw: JSONSchema7, name?: string, isRequired?: boolean): Schema {
   // TODO: validate that the raw JSON Schema is a valid JSON Schema before attempting to parse it.
 
+  if (raw.$ref) {
+    const __refName = raw.$ref.split('/').pop() || '';
```

Comment:
`__` is generally not a good practice to denote internal variables in javascript. Is there anyother special reason we are using `__refName` ?

Reply:
We’re using it here purely as an internal marker, typically during intermediate processing steps like schema resolution or codegen.
The idea is to:
    Avoid naming collisions with real schema fields or property names.
    Signal that it’s not meant to be persisted or exposed in the final output.